File: go/client_test.go
package postalcode

import (
	"context"
	"testing"
)

func TestListSearchAndHierarchy(t *testing.T) {
	client, err := NewClient()
	if err != nil {
		t.Fatal(err)
	}
	defer client.Close()
	regions, err := client.ListRegions(context.Background(), ListOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if len(regions) != 13 {
		t.Fatalf("got %d regions, want 13", len(regions))
	}
	attica, err := client.SearchRegions(context.Background(), "Αττικης", ListOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if len(attica) != 1 || attica[0].Name != "Αττικής" {
		t.Fatalf("unexpected search result: %#v", attica)
	}
	municipalities, err := client.SearchMunicipalities(context.Background(), "Αθην", MunicipalityOptions{ListOptions: ListOptions{IncludeHierarchy: true}})
	if err != nil {
		t.Fatal(err)
	}
	if municipalities[0].Hierarchy["regionalUnit"].Name != "Κεντρικού Τομέα Αθηνών" {
		t.Fatalf("unexpected hierarchy: %#v", municipalities[0].Hierarchy)
	}
}

func TestPostcodeAndAddressValidation(t *testing.T) {
	client, err := NewClient()
	if err != nil {
		t.Fatal(err)
	}
	defer client.Close()
	postcode, err := client.GetPostcode(context.Background(), "10431", PostcodeOptions{IncludeHierarchy: true, IncludeStreets: true})
	if err != nil {
		t.Fatal(err)
	}
	if postcode == nil || postcode.Hierarchy["municipality"].Name != "Αθηναίων" {
		t.Fatalf("unexpected postcode: %#v", postcode)
	}
	foundStreet := false
	for _, street := range postcode.Streets {
		foundStreet = foundStreet || street.Name == "Αγίου Κωνσταντίνου"
	}
	if !foundStreet {
		t.Fatal("expected Αγίου Κωνσταντίνου")
	}
	street := "Βενιζέλου Ελευθερίου"
	validation, err := client.ValidateAddress(context.Background(), AddressInput{Postcode: "10431", Street: &street, HouseNumber: 69, Municipality: "Αθην"})
	if err != nil {
		t.Fatal(err)
	}
	if validation.Postcode.Status != Valid || validation.HouseNumber.Status != Valid || validation.Municipality.Status != Valid {
		t.Fatalf("unexpected validation: %#v", validation)
	}
	missingStreet := "Δεν Υπάρχει"
	invalid, err := client.ValidateAddress(context.Background(), AddressInput{Postcode: "10431", Street: &missingStreet, HouseNumber: 1})
	if err != nil {
		t.Fatal(err)
	}
	if invalid.Street.Status != Invalid || invalid.HouseNumber.Status != NotEvaluated {
		t.Fatalf("unexpected invalid result: %#v", invalid)
	}
}

func TestClosePreventsQueries(t *testing.T) {
	client, err := NewClient()
	if err != nil {
		t.Fatal(err)
	}
	if err := client.Close(); err != nil {
		t.Fatal(err)
	}
	if _, err := client.ListRegions(context.Background(), ListOptions{}); err == nil {
		t.Fatal("expected closed-client error")
	}
}

File: dart/test/client_test.dart
import 'package:greek_postal_code_db/greek_postal_code_db.dart';
import 'package:test/test.dart';

void main() {
  test('core client contract', () async {
    final client = await createPostalCodeClient();
    try {
      if (client.listRegions().length != 13)
        throw StateError('Expected 13 regions.');
      if (client.searchRegions('Αττικης').single.name != 'Αττικής')
        throw StateError('Greek normalization failed.');
      final municipality =
          client.searchMunicipalities('Αθην', includeHierarchy: true).first;
      if (municipality.hierarchy?.regionalUnit?.name !=
          'Κεντρικού Τομέα Αθηνών')
        throw StateError('Entity hierarchy failed.');
      final postcode = client.getPostcode('10431',
          include: const PostcodeInclude(hierarchy: true, streets: true));
      if (postcode?.hierarchy?.municipality?.name != 'Αθηναίων' ||
          !(postcode!.streets!.any((s) => s.name == 'Αγίου Κωνσταντίνου')))
        throw StateError('Postcode lookup failed.');
      final valid = client.validateAddress(const AddressInput(
          postcode: '10431',
          street: 'Βενιζέλου Ελευθερίου',
          houseNumber: 69,
          municipality: LocalityName('Αθην')));
      if (valid.postcode.status != ValidationStatus.valid ||
          valid.street?.status != ValidationStatus.valid ||
          valid.houseNumber?.status != ValidationStatus.valid ||
          valid.municipality?.status != ValidationStatus.valid)
        throw StateError('Address validation failed.');
      final invalid = client.validateAddress(const AddressInput(
          postcode: '10431', street: 'Δεν Υπάρχει', houseNumber: 1));
      if (invalid.street?.status != ValidationStatus.invalid ||
          invalid.houseNumber?.status != ValidationStatus.notEvaluated)
        throw StateError('Dependent validation failed.');
    } finally {
      await client.close();
    }
  });
}
